src/mpi/comm/commutil.c

```diff
@@ -35,14 +35,122 @@ MPIR_Object_alloc_t MPIR_Comm_mem = {
 
 /* Communicator creation functions */
 struct MPIR_Commops *MPIR_Comm_fns = NULL;
-struct MPIR_Comm_hint_fn_elt {
-    char name[MPI_MAX_INFO_KEY];
+static int MPIR_Comm_commit_internal(MPIR_Comm * comm);
+
+/* Communicator hint functions */
+/* For balance of simplicity and feature, we'll internally use integers for both keys
+ * and values, and provide facilities to translate from and to string-based infos.
+ */
+
+struct MPIR_HINT {
+    const char *key;
     MPIR_Comm_hint_fn_t fn;
-    void *state;
-    UT_hash_handle hh;
+    int type;
+    int attr;                   /* e.g. whether this key is local */
 };
-static struct MPIR_Comm_hint_fn_elt *MPID_hint_fns = NULL;
-static int MPIR_Comm_commit_internal(MPIR_Comm * comm);
+static struct MPIR_HINT MPIR_comm_hint_list[MPIR_COMM_HINT_MAX];
+static int next_comm_hint_index = MPIR_COMM_HINT_PREDEFINED_COUNT;
+
+int MPIR_Comm_register_hint(int index, const char *hint_key, MPIR_Comm_Comm_hint_fn_t fn,
+                            int type, int attr)
```

Comment:
`fn` can be used to validate the hint value and possibly override the use suggested value, right? Can you add a code comment about that?

Reply:
Would it make sense to have this function (pointer) just validate the suggested value and return a new value if it is updated by the lower layers. Then we store that value in the hints table from here?

Reply:
> `fn` can be used to validate the hint value and possibly override the use suggested value, right? Can you add a code comment about that?

Sure.

Reply:
> Would it make sense to have this function (pointer) just validate the suggested value and return a new value if it is updated by the lower layers. Then we store that value in the hints table from here?

If a hook function is registered, the hook function is responsible for everything, include validation and setting the hints array, or updating other state variables wherever fits.